CHANGELOG.md

```diff
@@ -22,6 +22,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [466](https://github.com/sqlfluff/sqlfluff/issues/466) - Added L034
   'Fields should be stated before aggregates / window functions' per
   [dbt coding convenventions](https://github.com/fishtown-analytics/corp/blob/master/dbt_coding_conventions.md#sql-style-guide.)
+- Templating tags, such as `{{ variables }}`, `{# comments #}` and
+  `{% loops %}` (in jinja) now have placeholders in the parsed
+  structure. Rule L003 (indentation), also now respects these
+  placeholders so that their indentation is linted accordingly.
+  For loop or block tags, they also generate an `Indent` and
+  `Dedent` tag accordingly (which can be enabled or disabled)
+  with a configuration value so that indentation around these
+  functions can be linted accordingly.
```

Comment:
Are we running the linter twice? Once with the placeholders and once with the fully expanded SQL ?

Reply:
No. We only run it on the fully expanded SQL. The placeholders will be within the fully expanded sql, but they have zero length (i.e. their `.raw()` element is a zero length string `""`). This is very similar to how I've implemented indentation, where `Indent` and `Dedent` segments which also have zero length.

If you run `sqlfluff parse` on a basic file with some simple templating in it you'll see the parsed structure including any placeholders. That is the structure which is linted.

Reply:
Do you think the current docs explain that well? I wrote the docs in parallel with the code, and I'm often revising large patches of this code at the moment so it's very possible that the docs were well intentioned at the time but now need to be a bit clearer.

Reply:
This Changelog note is clear, however I'm thinking for a user coming from DBT trying to implement some rules that apply pre-templating, like the issues that @NiallRees has been opening (and I'm also interested in).

Does this work belong in a different PR? How do you think that would work? Would we end up with 2 ASTs? Would we have to tag issues to be "pre-templating" "post-templating" or "both" so they are selectively applied? <- Those are the ones keeping me up at night :)

Reply:
I've added some stubs into the code to open up the capability to fix certain things like the rules you've mentioned. There's a `TODO` note in the `linter.py`. But yes, this PR is already a monster. That is a job for another day.

If @NiallRees is watching 😄 : My intent is that a `TemplatedSegment` could have a `fixed_str` attribute, that when found would be used to replace the original version of the marker. It won't work in all cases - the way I've lexed templated code means that on very complicated jinja examples, I have to start making approximations, and in some of those we never get a `TemplatedSegment` in the lexed structure (so far it looks like I'm particularly bad at recognizing if statements).